runners/apex/src/main/java/org/apache/beam/runners/apex/translation/utils/ApexTimerInternals.java

```
@@ -0,0 +1,201 @@
+ */
+package org.apache.beam.runners.apex.translation.utils;
+
+import com.datatorrent.netlet.util.Slice;
+import com.esotericsoftware.kryo.DefaultSerializer;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import org.apache.beam.runners.core.StateNamespace;
+import org.apache.beam.runners.core.TimerInternals;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.joda.time.Instant;
+
+/**
+ * An implementation of Beam's {@link TimerInternals}.
+ *
+ * <p>Assumes that the current key is set prior to accessing the state.<br>
+ * This implementation stores timer data in heap memory and is serialized
+ * during checkpointing, it will only work with a small number of timers.
+ * @param <K>
+ */
+@DefaultSerializer(JavaSerializer.class)
+public class ApexTimerInternals<K> implements TimerInternals, Serializable {
+
+  private Map<Slice, Set<Slice>> activeTimers = new HashMap<>();
+  private TimerDataCoder timerDataCoder;
+  private transient K currentKey;
+  private transient Instant currentInputWatermark;
+  private transient Coder<K> keyCoder;
+
+  public ApexTimerInternals(TimerDataCoder timerDataCoder) {
+    this.timerDataCoder = timerDataCoder;
+  }
+
+  public void setContext(K key, Coder<K> keyCoder, Instant inputWatermark) {
+    this.currentKey = key;
+    this.keyCoder = keyCoder;
+    this.currentInputWatermark = inputWatermark;
+  }
+
+  @Override
+  public void setTimer(StateNamespace namespace, String timerId, Instant target,
+      TimeDomain timeDomain) {
+    TimerData timerData = TimerData.of(timerId, namespace, target, timeDomain);
```

I understand this is just a move of existing code. But here you enqueue timers for all time domains and below treat them all as event time timers. I think either you should make two sets/queues or else throw unsupported for processing time timers.

Is it correct that something like watermark update tuples will continue to cause the operator to inquire about ready timers? In that case the two queue solution is easy and it is actually OK to use `Instant.now()` since there's no distributed clock problem.